text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,483 @@
+
+In such situation the the client SHOULD implement queuing, acknowledgement handling and retrying logic per destination. This ensures that servers do not block each other. The queues SHOULD reference shared, immutable data to be sent, thus minimizing the memory overhead caused by having multiple queues.
+
+![Multi-Destination Exporting](images/otlp-multi-destination.png)
+
+This ensures that all destination servers receive the data regardless of their speed of reception (within the available limits imposed by the size of the client-side queue).
+
+### Load Balancing
+
+Implementations SHOULD utilize gRPC stream in a way that allows Level 7 load balancers that may be in the network path between the client and the server to re-balance the traffic periodically. This is necessary so that the traffic is not pinned by the load balancer to one server for the entire duration of telemetry data sending, thus potentially leading to imbalanced load of servers located behind the load balancer.
+
+Implementations can achieve this by periodically closing and reopening the gRPC stream. This will trigger the re-balancing by Level 7 load balancers that are gRPC-aware (such as Envoy) since they make re-balancing decisions when a new stream is opened.
+
+![Load Balancing](images/otlp-load-balancing-timeline.png)
+
+The client SHOULD periodically close and re-open the stream to help Load Balancers to re-balance the traffic.
+
+Client SHOULD close and re-open the stream after sending an `Export` request if any of these conditions is true:
+
+- Time passed since the stream was last opened is longer than a user-configurable re-balancing period `RebalanceTimeout`.
+
+- Number of requests exported since the stream was last opened is greater than user-configurable request count limit `RebalanceRequests`.
+
+In Pipelined mode the client SHOULD continue processing responses that the server may send to a stream that is closed from the client side until that stream is closed from the server side as well. This means that during this transition there may exist more than one stream of data between the client and the server.
+
+Similar stream re-opening MAY also be initiated from the server-side, however it is usually less preferable since it results in the loss of requests which are not yet acknowledged and requires the client to re-send the requests. Even if the client is well-behaved and does the re-sending correctly this still results in overall worse throughput of the protocol. Implementations where the server needs to serve a mix of clients some of which are utilizing stream re-opening strategy and some others don't are recommended to be configured `RebalanceTimeout` and `RebalanceRequests` values that are larger at server-side than the corresponding values at client-side. This way the protocol will fall back to stream re-opening at the server-side only for the clients, which don't already utilize client-side stream re-opening.
```

Comment:
It is implicit that this is not the default gRPC server side keepalive settings, perhaps it is worth mentioning that and their interaction with keepalive settings (they have defaults even if not specified). A simplifying assumption to implement these on the server side is that it receives each individual requests, respond to the request, and  then checks for `RebalanceTimeout` and `RebalanceRequests` thresholds.